---
title: "10092023_ZGA_List_intersection"
output:
  html_document:
    df_print: paged
---

# Title
# Date 10092023
# Name ZGA_List_intersection
# Description The R script for DE analysis and Pan ZGA genes list reproduction


## Loading libraries
```{r}
library(rhdf5)
library(conflicted)
conflict_prefer("select", "AnnotationDbi")
library(tximportData)
library(tximport)
library(withr)
library(base)
library(TxDb.Mmusculus.UCSC.mm10.knownGene) 
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(cluster)
library(dendextend)
library(circlize)
library(RColorBrewer)
library(ComplexHeatmap)
library(ggplot2)
library(stringr)
library(dplyr)
library(tidyverse)
library(biomaRt)
library(readr)
library(tibble)
library(plot3D)
#library(scBatch)
library(splatter)
#library(scuttle)
#library(scater)
library(rgl)
library(SingleCellExperiment)
library(scRNAseq)
library(biomaRt)
library(limma)
#library(edgeR)
library(umap)
library(tidyverse)
library(plot3D)
library(pheatmap)
library(DESeq2)
library(stringr)
library(assertive.numbers)
library(rhdf5)
library(tximport)
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(stringr)
library(ComplexHeatmap)
library(umap)
library(dplyr)
library(viridis)     
library(RColorBrewer)
library(scico)
library(DESeq2)
library(wesanderson)
library(data.table)
library(biomaRt)
library(edgeR)
library(ggplot2)
library(ggrepel)

"%+%" <- function(...){
  paste0(...)
}

```

```{r}
library(conflicted)
library(rhdf5)
library(org.Mm.eg.db)
library(tximport)
library(DESeq2)
library(dplyr)
library(ggplot2)
library(ggrepel)

"%+%" <- function(...){
  paste0(...)
}

source("./readKallisto.R")

# system("wget https://github.com/pachterlab/kallisto-transcriptome-indices/releases/download/ensembl-96/mus_musculus.tar.gz")
# system("tar -xf mus_musculus.tar.gz")

```


```{r}

# processRNA_Seq function: from a path to Kallisto files to a sumary DE table and volcano plot

processRNA_Seq <- function(path, directory, gtf.gr, plot_volcanoplot=T) {
  
  # solving some conflicts
  conflict_prefer("mutate", "dplyr")
  conflict_prefer("strsplit", "Biostrings")
  conflict_prefer("getSequence", "biomaRt")
  conflict_prefer("select", "dplyr")
  
  # setting up environment
  working_dir = path %+% "/" %+% directory
  
  # loading methadata
  samples_methadata <- read.csv(file.path(working_dir, "samples.csv"), header = TRUE, sep=";")
  
  # loading annotation
  gtf.df = as.data.frame(gtf.gr)
  txt2gene=(gtf.df[gtf.df$type=='transcript' ,c("transcript_id","gene_name")])
  names(txt2gene)<-c('TXNAME','GENEID')
  
  # loading kalisto h5
  h5closeAll() # close to avoid conflicts
  kallisto_files <- file.path(working_dir, "Kallisto", samples_methadata$file_path, "abundance.h5")
  names(kallisto_files) <- samples_methadata$file_path
  txi.kallisto <- tximport(kallisto_files, type = "kallisto", txOut = F, tx2gene = txt2gene, ignoreTxVersion = TRUE)
  
  write.table(txi.kallisto$counts,file=paste0(working_dir, "/txi.kallisto_counts_raw_genes.csv"),row.names=T,quote=F,sep="\t")
  
  # running DESeq2
  ddsSalmon <- DESeqDataSetFromTximport(txi.kallisto, colData = samples_methadata, design = ~condition)
  ddsSalmon <- DESeq(ddsSalmon)
  fpkmdds = as.data.frame(fpkm(ddsSalmon))
  fpkmdds$Zygote = apply(fpkmdds[, samples_methadata$sample_name[samples_methadata["condition"] == "Zygote"]],1,mean)
  fpkmdds$`2cell` = apply(fpkmdds[, samples_methadata$sample_name[samples_methadata["condition"] == "2cell"]],1,mean)
  fpkmdds = fpkmdds %>% select(c("Zygote", "2cell"))
  colnames(fpkmdds) = c("Zygote_FPKM", "2cell_FPKM")
  abdds = as.data.frame(txi.kallisto$abundance)
  abdds$Zygote = apply(abdds[, samples_methadata$sample_name[samples_methadata["condition"] == "Zygote"]],1,mean)
  abdds$`2cell` = apply(abdds[, samples_methadata$sample_name[samples_methadata["condition"] == "2cell"]],1,mean)
  abdds = abdds %>% dplyr::select(c("Zygote", "2cell"))
  colnames(abdds) = c("Zygote_Ab", "2cell_Ab")
  
  # importing abundances
  files <- dir(working_dir, "abundance.tsv", full=TRUE,
               recursive=TRUE)
  stopifnot(all(file.exists(files)))
  files <- sub(".tsv", ".h5", files, fixed=TRUE)
  exp = readKallisto(files, what=c("tpm", "eff_length"), as="SummarizedExperiment")
  tpmdds = as.data.frame(exp@assays@data$tpm)
  tpmdds$Zygote = apply(tpmdds[, samples_methadata$sample_name[samples_methadata["condition"] == "Zygote"]],1,mean)
  tpmdds$`2cell` = apply(tpmdds[, samples_methadata$sample_name[samples_methadata["condition"] == "2cell"]],1,mean)
  tpmdds = tpmdds %>% dplyr::select(c("Zygote", "2cell"))
  colnames(tpmdds) = c("Zygote_TPM", "2cell_TPM")
  tpmdds$ENTREZID = rownames(tpmdds) 
  tpmdds = tidyr::separate(tpmdds, col="ENTREZID", into = c("ENTREZID", "ENTREZID_dot")) %>% dplyr::select(c("ENTREZID", "Zygote_TPM", "2cell_TPM"))
  
  # merging dfs to form an output
  tpmdds <- merge(gtf.df %>% dplyr::select(c("transcript_id", "gene_name")),tpmdds,
                        by.x=1,
                        by.y=1,
                        all.x=FALSE,
                        all.y=TRUE)
  
  tpmdds = tpmdds[!duplicated(tpmdds), ]
  myResS <- results(ddsSalmon, contrast = c("condition", "Zygote", "2cell"))
  myResS <- myResS[order(myResS$pvalue), ]
  myResAsDF <- as.data.frame(myResS)
  myResAsDF$newPadj <- p.adjust(myResAsDF$pvalue, method ="fdr")
  
  conflict_prefer("select", "AnnotationDbi")
  eToSym <- select(org.Mm.eg.db,
                   keys = rownames(myResAsDF),
                   keytype = "SYMBOL",
                   columns=c("SYMBOL","ENTREZID","GENENAME"))
  
  annotatedRes <- merge(eToSym,myResAsDF,
                        by.x=1,
                        by.y=0,
                        all.x=TRUE,
                        all.y=TRUE)
  
  annotatedRes <- annotatedRes[order(annotatedRes$pvalue),]
  
  annotatedRes_fpkmdds <- merge(annotatedRes,fpkmdds,
                        by.x=1,
                        by.y=0,
                        all.x=TRUE,
                        all.y=FALSE)
  
  annotatedRes_fpkmdds <- merge(annotatedRes_fpkmdds,tpmdds,
                        by.x=1,
                        by.y=2,
                        all.x=TRUE,
                        all.y=FALSE)
  
  write.table(annotatedRes_fpkmdds,file=paste0(working_dir, "/annotatedRes_log2FC_" %+% directory %+% "_genes.csv"),row.names=T,quote=F,sep="\t")
  
  if (plot_volcanoplot == T) {
    
    # Visualizing DE genes on volcano plot
    annotatedRes %>% 
      mutate(padj_neg10 = -log10(padj)) -> res_mm
    
    FLAG = 4
    groupB_criteria <- res_mm$padj < 0.01 & abs(res_mm$log2FoldChange) > FLAG
    
    res_mm$significance <- "n.s."
    group1 = "padj < 0.01"
    res_mm$significance[res_mm$padj < 0.01] <- group1
    group2 = "padj < 0.01 & log2FoldChange > " %+% FLAG
    res_mm$significance[groupB_criteria] <- group2
    
    ggplot(res_mm, aes(log2FoldChange, padj_neg10, color = significance)) +
      geom_point(shape = 16, alpha = 0.6) +
      labs(title = "DESeq2 transcripts " %+% "Zygote_G2" %+% " vs " %+% "2cell_G2") +
      xlab("log2FoldChange")+ ylab("-log10(padj)")+
      guides(text = guide_legend()) +
      scale_color_manual(values = c("n.s." = "darkgray",
                                    "padj < 0.01" = "gray",
                                    "padj < 0.01 & log2FoldChange > 4" = "orange")) +
      NULL + theme_light()
    
    ggsave("DE_"%+% directory %+%".pdf")
  }
}
```


```{r}
home = "/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko"

#gtf.gr = rtracklayer::import("./mus_musculus/Mus_musculus.GRCm38.96.gtf")

processRNA_Seq(home, "ZGA_Gassler_C57BL_6J", gtf.gr)
processRNA_Seq(home, "ZGA_Gassler_B6xCAST", gtf.gr)
processRNA_Seq(home, "ZGA_Hu_from_Xie", gtf.gr)
processRNA_Seq(home, "ZGA_Obox_Shuyan", gtf.gr)
processRNA_Seq(home, "ZGA_Yu", gtf.gr)
processRNA_Seq(home, "ZGA_Wang", gtf.gr)
processRNA_Seq(home, "ZGA_Wu", gtf.gr)
processRNA_Seq(home, "ZGA_Zhang", gtf.gr)
processRNA_Seq(home, "ZGA_Dux_KO_Chen_single", gtf.gr)
processRNA_Seq(home, "ZGA_Deng_single", gtf.gr)
processRNA_Seq(home, "ZGA_Guo", gtf.gr)
processRNA_Seq(home, "ZGA_Zhang_2016", gtf.gr)
```


```{r}
library(dplyr)
library(magrittr)
conflict_prefer("select", "dplyr")

v_1cell_threshold = 2
v_2cell_threshold = 5
v_pval_threshold = 0.05
log2FoldChange_threshold = 2
home = "/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko"


read_and_threshold_ZGA_genes <- function(path, directory, v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold){
  working_dir = path %+% "/" %+% directory
  annotatedRes = read.csv(file=paste0(working_dir, "/annotatedRes_log2FC_" %+% directory %+% "_genes.csv"),sep="\t")
  annotatedRes %<>% dplyr::filter((log2FoldChange < log2FoldChange_threshold) & (newPadj < v_pval_threshold) & (Zygote_FPKM > v_1cell_threshold) & (X2cell_FPKM > v_2cell_threshold))
  return(annotatedRes)
}


annotatedRes_C57XBL_6J = read_and_threshold_ZGA_genes(home, "ZGA_Gassler_C57BL_6J", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_B6xCAST = read_and_threshold_ZGA_genes(home, "ZGA_Gassler_B6xCAST", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_C57XBL_6J = read_and_threshold_ZGA_genes(home, "ZGA_Hu_from_Xie", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_C57XBL_6J = read_and_threshold_ZGA_genes(home, "ZGA_Obox_Shuyan", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_C57XBL_6J = read_and_threshold_ZGA_genes(home, "ZGA_Yu", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_C57XBL_6J = read_and_threshold_ZGA_genes(home, "ZGA_Wang", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_C57XBL_6J = read_and_threshold_ZGA_genes(home, "ZGA_Wu", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_C57XBL_6J = read_and_threshold_ZGA_genes(home, "ZGA_Zhang", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_C57XBL_6J = read_and_threshold_ZGA_genes(home, "ZGA_Dux_KO_Chen_single", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_C57XBL_6J = read_and_threshold_ZGA_genes(home, "ZGA_Deng_single", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_C57XBL_6J = read_and_threshold_ZGA_genes(home, "ZGA_Guo", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_C57XBL_6J = read_and_threshold_ZGA_genes(home, "ZGA_Zhang_2016", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)


gene_list <- list(
  "B6xCAST Gassler" = unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
  "C57BL_6J Gassler" = unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),
  "C57BL_6N Hu from Xie" = unique(toupper(annotatedRes_Hu$SYMBOL)),
  "C57BL_6 Shuyan" = unique(toupper(annotatedRes_Shuyan$SYMBOL)),
  "ICR Yu" = unique(toupper(annotatedRes_Yu$SYMBOL)),
  "B6D2F1_J Wang" = unique(toupper(annotatedRes_Wang$SYMBOL)),
  "C57BL_6NxDBA_2N Wu" = unique(toupper(annotatedRes_Wu$SYMBOL)),
  "B6D2F1 Chen" = unique(toupper(annotatedRes_Chen$SYMBOL)),
  "CAST_EiJxC57BL_6J Deng" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "B6D2F1_J Zhang" = unique(toupper(annotatedRes_Zhang$SYMBOL)),
  "C57BL_6 Guo" = unique(toupper(annotatedRes_Guo$SYMBOL)))


G_ZGA = Reduce(intersect,list(unique(toupper(annotatedRes_B6xCAST$SYMBOL)),unique(toupper(annotatedRes_C57XBL_6J$SYMBOL))))

gene_list2 <- list(
  "ZGA Gassler" = G_ZGA,
  "C57BL_6N Hu from Xie" = unique(toupper(annotatedRes_Hu$SYMBOL)),
  "C57BL_6 Shuyan" = unique(toupper(annotatedRes_Shuyan$SYMBOL)),
  "C57BL_6 Yu" = unique(toupper(annotatedRes_Yu$SYMBOL)),
  "B6D2F1_J Wang" = unique(toupper(annotatedRes_Wang$SYMBOL)),
  "C57BL_6NxDBA_2N Wu" = unique(toupper(annotatedRes_Wu$SYMBOL)),
  "B6D2F1 Chen" = unique(toupper(annotatedRes_Chen$SYMBOL)),
  "CAST_EiJ(mother)xC57BL_6J(father) Deng" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "B6D2F1_J Zhang" = unique(toupper(annotatedRes_Zhang$SYMBOL)))

gene_list_polyA <- list(
  "B6xCAST Gassler" = unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
  "C57BL_6J Gassler" = unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),
  "C57BL_6N Hu from Xie" = unique(toupper(annotatedRes_Hu$SYMBOL)),
  "C57BL_6 Shuyan" = unique(toupper(annotatedRes_Shuyan$SYMBOL)),
  "C57BL_6 Yu" = unique(toupper(annotatedRes_Yu$SYMBOL)),
  "C57BL_6NxDBA_2N Wu" = unique(toupper(annotatedRes_Wu$SYMBOL)),
  "B6D2F1 Chen" = unique(toupper(annotatedRes_Chen$SYMBOL)),
  "CAST_EiJxC57BL_6J Deng" = unique(toupper(annotatedRes_Deng$SYMBOL)))


Reduce(intersect,gene_list_polyA)[[71]]

library(eulerr)
p <- plot(euler(gene_list, shape="ellipse"), quantities = TRUE, legend = TRUE)
print(p)

library(nVennR)

nVennR_gene_list_polyA_total <- list(
  "C57BL (5)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Hu$SYMBOL)), unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_Yu$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),unique(toupper(annotatedRes_Guo$SYMBOL)) ))),
  "B6xCAST (1)" = unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
  "CAST/EiJ(mother) x C57BL/6J(father) (1)" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "B6D2F1 (3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Chen$SYMBOL)), unique(toupper(annotatedRes_Wang$SYMBOL)), unique(toupper(annotatedRes_Zhang$SYMBOL))))),
  "C57BL_6NxDBA_2N (1)" = unique(toupper(annotatedRes_Wu$SYMBOL))
  )

nVennR_gene_list_polyA_total <- list(
  "C57BL (4)" = unique(Reduce(intersect,list(unique(toupper(annotatedRes_Hu$SYMBOL)), unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_Yu$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL))))),
  "B6xCAST (1)" = unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
  "CAST/EiJ(mother) x C57BL/6J(father) (1)" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "B6D2F1 (3)" = unique(Reduce(intersect,list(unique(toupper(annotatedRes_Chen$SYMBOL)), unique(toupper(annotatedRes_Wang$SYMBOL)), unique(toupper(annotatedRes_Zhang$SYMBOL))))),
  "C57BL_6NxDBA_2N (1)" = unique(toupper(annotatedRes_Wu$SYMBOL))
  )

nVennR_gene_list_polyA <- list(
  "C57BL (4)" = unique(Reduce(intersect,list(unique(toupper(annotatedRes_Hu$SYMBOL)), unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_Yu$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL))))),
  "B6xCAST (1)" = unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
  "CAST/EiJ(mother) x C57BL/6J(father) (1)" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "B6D2F1 (1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
  "C57BL_6NxDBA_2N (1)" = unique(toupper(annotatedRes_Wu$SYMBOL))
  )

nVennR_gene_list_polyA <- list(
  "C57BL (3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),unique(toupper(annotatedRes_Guo$SYMBOL))))),
  "B6xCAST (1)" = unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
  "CAST/EiJ x C57BL/6J (1)" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "B6D2F1 (1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
  "C57BL_6NxDBA_2N (1)" = unique(toupper(annotatedRes_Wu$SYMBOL)),
    "ICR" = unique(toupper(annotatedRes_Yu$SYMBOL))
  )

nVennR_gene_list_polyA <- list(
  "C57BL/6J (3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),unique(toupper(annotatedRes_Guo$SYMBOL))))),
  "C57BL/6JxCAST/EiJ (1)" = unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
  "CAST/EiJxC57BL/6J (1)" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "B6D2F1 (1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
  "C57BL/6NxDBA/2N (1)" = unique(toupper(annotatedRes_Wu$SYMBOL)),
  "ICR (1)" = unique(toupper(annotatedRes_Yu$SYMBOL)),
  "C57BL/6NxPWK/PhJ (1)" = unique(toupper(annotatedRes_Zhang_2016$SYMBOL))
  )

nVennR_gene_list_polyA <- list(
  "C57BL/6J (3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),unique(toupper(annotatedRes_Guo$SYMBOL))))),
  "C57BL/6JxCAST/EiJ (2)" = unique(Reduce(intersect,list(unique(toupper(annotatedRes_B6xCAST$SYMBOL)),unique(toupper(annotatedRes_Deng$SYMBOL))))),
  "B6D2F1 (1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
  "C57BL/6NxDBA/2N (1)" = unique(toupper(annotatedRes_Wu$SYMBOL)),
  "ICR (1)" = unique(toupper(annotatedRes_Yu$SYMBOL)),
  "C57BL/6NxPWK/PhJ (1)" = unique(toupper(annotatedRes_Zhang_2016$SYMBOL))
  )


nVennR_gene_list_polyA <- list(
  "C57BL/6J clean (N=3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),unique(toupper(annotatedRes_Guo$SYMBOL))))),
  "C57BL/6J x mix (N=3)" = unique(Reduce(intersect,list(unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
                                                 unique(toupper(annotatedRes_Wu$SYMBOL)),
                                              unique(toupper(annotatedRes_Zhang_2016$SYMBOL))))),
  "CAST/EiJxC57BL/6J (N=1)" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "B6D2F1 clean (N=1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
  "ICR clean (N=1)" = unique(toupper(annotatedRes_Yu$SYMBOL))
  )



nVennR_gene_list_polyA <- list(
    "C57BL/6NxPWK/PhJ (1)" = unique(toupper(annotatedRes_Zhang_2016$SYMBOL)),
      "C57BL/6JxCAST/EiJ (1)" = unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
  "C57BL/6NxDBA/2N (1)" = unique(toupper(annotatedRes_Wu$SYMBOL)),
    "ICR (1)" = unique(toupper(annotatedRes_Yu$SYMBOL)),
  "C57BL/6J (3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),unique(toupper(annotatedRes_Guo$SYMBOL)))))
  )


nVennR_gene_list_polyA <- list(
    
     "B6D2F1 clean (N=1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
     "ICR clean (1)" = unique(toupper(annotatedRes_Yu$SYMBOL)),
     "C57BL/6J x mix (N=3)" = unique(Reduce(intersect,list(unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
                                                 unique(toupper(annotatedRes_Wu$SYMBOL)),
                                              unique(toupper(annotatedRes_Zhang_2016$SYMBOL))))),
     "CAST/EiJxC57BL/6J (N=1)" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "C57BL/6J clean (3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),unique(toupper(annotatedRes_Guo$SYMBOL)))))
  )
myV <- plotVenn(nVennR_gene_list_polyA, outFile="/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/plotVenn_polyA_1.svg")

nVennR_gene_list_polyA <- list(
    "ICR clean (1)" = unique(toupper(annotatedRes_Yu$SYMBOL)),
     "B6D2F1 clean (N=1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
     
     "C57BL/6J x mix (N=3)" = unique(Reduce(intersect,list(unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
                                                 unique(toupper(annotatedRes_Wu$SYMBOL)),
                                              unique(toupper(annotatedRes_Zhang_2016$SYMBOL))))),
     "CAST/EiJxC57BL/6J (N=1)" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "C57BL/6J clean (3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),unique(toupper(annotatedRes_Guo$SYMBOL)))))
  )
myV <- plotVenn(nVennR_gene_list_polyA, outFile="/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/plotVenn_polyA_2.svg")

nVennR_gene_list_polyA <- list(
    
     "B6D2F1 clean (N=1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
     "ICR clean (1)" = unique(toupper(annotatedRes_Yu$SYMBOL)),
     "C57BL/6J x mix (N=3)" = unique(Reduce(intersect,list(unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
                                                 unique(toupper(annotatedRes_Wu$SYMBOL)),
                                              unique(toupper(annotatedRes_Zhang_2016$SYMBOL))))),
  "C57BL/6J clean (3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),unique(toupper(annotatedRes_Guo$SYMBOL))))),
                                       "CAST/EiJxC57BL/6J (N=1)" = unique(toupper(annotatedRes_Deng$SYMBOL))
  )
myV <- plotVenn(nVennR_gene_list_polyA, outFile="/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/plotVenn_polyA_3.svg")



myV <- plotVenn(nVennR_gene_list_polyA_total, outFile="/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/plotVenn_polyA_total.svg")
myV <- plotVenn(nVennR_gene_list_polyA, outFile="/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/plotVenn_polyA.svg")
myV <- plotVenn(nVennR_gene_list_polyA_total, outFile="/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/plotVenn_polyA_4.svg")
myV <- plotVenn(nVennR_gene_list_polyA, outFile="/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/plotVenn_polyA_4.svg")


library(UpSetR)

set_size = function(w, h, factor=1.5) {
    s = 1 * factor
    options(
        repr.plot.width=w * s,
        repr.plot.height=h * s,
        repr.plot.res=100 / factor,
        jupyter.plot_mimetypes='image/png',
        jupyter.plot_scale=1
    )
}

library(ggplot2)
library(ComplexUpset)


set_size(8, 5)
size = get_size_mode('exclusive_intersection')
ComplexUpset::upset(fromList(gene_list)[apply(fromList(gene_list),1,sum) > 8, ], 
                    names(gene_list), 
                    #intersections='all', 
                    width_ratio=0.1, 
                    #min_size=50,
                    #max_size=400,
                    base_annotations = list(
        'Intersection size'=(
            intersection_size(text=list(
                vjust=-0.1,
                hjust=-0.1,
                angle=45,
                color="black"),
                
                text_mapping=aes(
                #label=paste0(
                #    !!upset_text_percentage(),
                #    '\n(', !!size, ')'
                #),
                colour=ifelse(!!size > 50, 'on_bar', 'on_background'),
                y=ifelse(!!size > 50, !!size, !!size)
            )
            
                )
            + ylim(c(0, 160))
            + theme(plot.background=element_rect(fill='#F2F2E1'),axis.text=element_text(size=12),axis.title=element_text(size=14))
            + ylab('Overlapped genes')
        )
    ),
    #height_ratio=1,
    sort_sets='descending',
    set_sizes=(upset_set_size(filter_intersections=TRUE) + scale_y_continuous(
            trans=reverse_log_trans(),
            labels=log10
        )
        + ylab('log10(set size)')
        
        ),
    stripes='white', 
    sort_intersections_by=c('degree', 'cardinality')
        )
              

ComplexUpset::upset(fromList(gene_list_polyA)[apply(fromList(gene_list_polyA),1,sum) > 6, ], 
                    names(gene_list_polyA), 
                    #intersections='all', 
                    width_ratio=0.1, 
                    #min_size=50,
                    #max_size=400,
                    base_annotations = list(
        'Intersection size'=(
            intersection_size(text=list(
                vjust=-0.1,
                hjust=-0.1,
                angle=45,
                color="black"),
                
                text_mapping=aes(
                #label=paste0(
                #    !!upset_text_percentage(),
                #    '\n(', !!size, ')'
                #),
                colour=ifelse(!!size > 50, 'on_bar', 'on_background'),
                y=ifelse(!!size > 50, !!size, !!size)
            )
            
                )
            + ylim(c(0, 200))
            + theme(axis.text=element_text(size=12),axis.title=element_text(size=14), 
                    axis.line = element_line(colour = "black"),
                    panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),axis.ticks.x=element_line(),axis.ticks.y=element_line()) 
            + ylab('Overlapped genes')
        )
    ),
    #height_ratio=1,
    sort_sets='descending',
    set_sizes=(upset_set_size() + theme(axis.text.x=element_text(angle=90))
        + ylab('set size')
        
        ),
    stripes='white', 
    sort_intersections_by=c('degree', 'cardinality')
        ) 



ComplexUpset::upset(fromList(gene_list)[apply(fromList(gene_list),1,sum) > 8, ], 
                    names(gene_list), 
                    #intersections='all', 
                    width_ratio=0.1, 
                    #min_size=50,
                    #max_size=400,
                    base_annotations = list(
        'Intersection size'=(
            intersection_size(text=list(
                vjust=-0.1,
                hjust=-0.1,
                angle=45,
                color="black"),
                
                text_mapping=aes(
                #label=paste0(
                #    !!upset_text_percentage(),
                #    '\n(', !!size, ')'
                #),
                colour=ifelse(!!size > 50, 'on_bar', 'on_background'),
                y=ifelse(!!size > 50, !!size, !!size)
            )
            
                )
            + ylim(c(0, 160))
            + theme(axis.text=element_text(size=12),axis.title=element_text(size=14), 
                    axis.line = element_line(colour = "black"),
                    panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),axis.ticks.x=element_line(),axis.ticks.y=element_line()) 
            + ylab('Overlapped genes')
        )
    ),
    #height_ratio=1,
    sort_sets='descending',
    set_sizes=(upset_set_size() + theme(axis.text.x=element_text(angle=90))
        + ylab('set size')
        
        ),
    stripes='white', 
    sort_intersections_by=c('degree', 'cardinality')
        ) 


ComplexUpset::upset(fromList(nVennR_gene_list_polyA)[apply(fromList(nVennR_gene_list_polyA),1,sum) > 3, ], 
                    names(nVennR_gene_list_polyA), 
                    #intersections='all', 
                    width_ratio=0.1, 
                    #min_size=50,
                    #max_size=400,
                    base_annotations = list(
        'Intersection size'=(
            intersection_size(text=list(
                vjust=-0.1,
                hjust=-0.1,
                angle=45,
                color="black"),
                
                text_mapping=aes(
                #label=paste0(
                #    !!upset_text_percentage(),
                #    '\n(', !!size, ')'
                #),
                colour=ifelse(!!size > 50, 'on_bar', 'on_background'),
                y=ifelse(!!size > 50, !!size, !!size)
            )
            
                )
            + ylim(c(0, 1100))
            + theme(axis.text=element_text(size=12),axis.title=element_text(size=14), 
                    axis.line = element_line(colour = "black"),
                    panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),axis.ticks.x=element_line(),axis.ticks.y=element_line()) 
            + ylab('Overlapped genes')
        )
    ),
    #height_ratio=1,
    sort_sets='descending',
    set_sizes=(upset_set_size() + theme(axis.text.x=element_text(angle=90))
        + ylab('set size')
        
        ),
    stripes='white', 
    sort_intersections_by=c('degree', 'cardinality')
        ) 


set_size(8, 5)
size = get_size_mode('exclusive_intersection')
ComplexUpset::upset(fromList(nVennR_gene_list_polyA_total), #[apply(fromList(nVennR_gene_list_polyA_total),1,sum) > 3, ], 
                    names(nVennR_gene_list_polyA_total), 
                    #intersections='all', 
                    width_ratio=0.1, 
                    #min_size=50,
                    #max_size=400,
                    base_annotations = list(
        'Intersection size'=(
            intersection_size(text=list(
                vjust=-0.1,
                hjust=-0.1,
                angle=45,
                color="black"),
                
                text_mapping=aes(
                #label=paste0(
                #    !!upset_text_percentage(),
                #    '\n(', !!size, ')'
                #),
                colour=ifelse(!!size > 50, 'on_bar', 'on_background'),
                y=ifelse(!!size > 50, !!size, !!size)
            )
            
                )
            + ylim(c(0, 1750))
            + theme(axis.text=element_text(size=12),axis.title=element_text(size=14), 
                    axis.line = element_line(colour = "black"),
                    panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),axis.ticks.x=element_line(),axis.ticks.y=element_line()) 
            + ylab('Overlapped genes')
        )
    ),
    #height_ratio=1,
    sort_sets='descending',
    set_sizes=(upset_set_size() + theme(axis.text.x=element_text(angle=90))
        + ylab('set size')
        
        ),
    stripes='white', 
    sort_intersections_by=c('degree', 'cardinality')
        ) 

#+ theme(axis.text.x=element_text(angle=90),axis.ticks.x=element_line())+scale_y_continuous(trans=reverse_log_trans()))



```



```{r}


cc = c(nrow(fromList(gene_list)[apply(fromList(gene_list),1,sum) == 10, ]))

for (i in seq(length(gene_list))){
  print(i)
  cc = c(cc, nrow(fromList(gene_list[- i])[apply(fromList(gene_list[- i]),1,sum) == 9, ]))
}

cc
cn = c("All", paste0("-",names(gene_list)))
cn
df = as.data.frame(list(Overlap=cc,Datasets=cn))

range = max(df$Overlap) -min(df$Overlap)

ggplot(df, aes(x=fct_inorder(cn), y=cc))+
geom_bar(stat="identity",colour=NA,linewidth=0)+
  theme_classic()+
  ylab("Overlaped genes")+
  xlab("Datasets")+theme(axis.text.x = element_text(angle = 90),
                                                                 axis.line = element_line(colour = "black"),
                                                                 text = element_text(size = 12),
                                                                 panel.grid.major = element_blank(),
  panel.grid.minor = element_blank())+
  geom_text(aes(label = cc), 
            angle = 45,
            nudge_y = range/3)



```



```{r}


cc = c(nrow(fromList(gene_list_polyA)[apply(fromList(gene_list_polyA),1,sum) == 8, ]))

for (i in seq(length(gene_list_polyA))){
  print(i)
  cc = c(cc, nrow(fromList(gene_list_polyA[- i])[apply(fromList(gene_list_polyA[- i]),1,sum) == 7, ]))
}

cc
cn = c("All", paste0("-",names(gene_list_polyA)))
cn
df = as.data.frame(list(Overlap=cc,Datasets=cn))

range = max(df$Overlap) -min(df$Overlap)

ggplot(df, aes(x=fct_inorder(cn), y=cc))+
geom_bar(stat="identity",colour=NA,linewidth=0)+
  theme_classic()+
  ylab("Overlaped genes")+
  xlab("Datasets")+theme(axis.text.x = element_text(angle = 90),
                                                                 axis.line = element_line(colour = "black"),
                                                                 text = element_text(size = 12),
                                                                 panel.grid.major = element_blank(),
  panel.grid.minor = element_blank())+
  geom_text(aes(label = cc), 
            angle = 45,
            nudge_y = range/3)



```


```{r}


cc = c(nrow(fromList(nVennR_gene_list_polyA)[apply(fromList(nVennR_gene_list_polyA),1,sum) == 5, ]))

for (i in seq(length(nVennR_gene_list_polyA))){
  print(i)
  cc = c(cc, nrow(fromList(nVennR_gene_list_polyA[- i])[apply(fromList(nVennR_gene_list_polyA[- i]),1,sum) == 4, ]))
}

cc
cn = c("All", paste0("-",names(nVennR_gene_list_polyA)))
cn
df = as.data.frame(list(Overlap=cc,Datasets=cn))

range = max(df$Overlap) -min(df$Overlap)

ggplot(df, aes(x=fct_inorder(cn), y=cc))+
geom_bar(stat="identity",colour=NA,linewidth=0)+
  theme_classic()+
  ylab("Overlaped genes")+
  xlab("Datasets")+theme(axis.text.x = element_text(angle = 90),
                                                                 axis.line = element_line(colour = "black"),
                                                                 text = element_text(size = 12),
                                                                 panel.grid.major = element_blank(),
  panel.grid.minor = element_blank())+
  geom_text(aes(label = cc), 
            angle = 45,
            nudge_y = range/3)



```
```{r}


cc = c(nrow(fromList(nVennR_gene_list_polyA_total)[apply(fromList(nVennR_gene_list_polyA_total),1,sum) == 5, ]))

for (i in seq(length(nVennR_gene_list_polyA_total))){
  print(i)
  cc = c(cc, nrow(fromList(nVennR_gene_list_polyA_total[- i])[apply(fromList(nVennR_gene_list_polyA_total[- i]),1,sum) == 4, ]))
}

cc
cn = c("All", paste0("-",names(nVennR_gene_list_polyA_total)))
cn
df = as.data.frame(list(Overlap=cc,Datasets=cn))

range = max(df$Overlap) -min(df$Overlap)

ggplot(df, aes(x=fct_inorder(cn), y=cc))+
geom_bar(stat="identity",colour=NA,linewidth=0)+
  theme_classic()+
  ylab("Overlaped genes")+
  xlab("Datasets")+theme(axis.text.x = element_text(angle = 90),
                                                                 axis.line = element_line(colour = "black"),
                                                                 text = element_text(size = 12),
                                                                 panel.grid.major = element_blank(),
  panel.grid.minor = element_blank())+
  geom_text(aes(label = cc), 
            angle = 45,
            nudge_y = range/3)



```



```{r}
suppressPackageStartupMessages({
  library(tidyverse)
  library(readxl)
  library(DESeq2)
  library(eulerr)
})

ZGA_Chen_et_al <- file.path("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/Chen_et_al_RNA_seq/41588_2019_418_MOESM3_ESM.xlsx")
ZGA_Chen_et_al <- read_xlsx(ZGA_Chen_et_al, sheet=2)
ZGA_Chen_et_al = ZGA_Chen_et_al[which(ZGA_Chen_et_al$...17 == "down-regulated"),]
Chen_et_al_Dux_targets = ZGA_Chen_et_al[["...1"]] #[-c(1,2)]

res_mm = read.delim("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/science.abn7478_tables_s1_to_s14/science.abn7478_table_s6.csv", sep = ";")
nr5a2_kd = unique(res_mm[which(res_mm$State == "DOWN"),]$gene_id)

res_mm = read.delim("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/science.abn7478_tables_s1_to_s14/science.abn7478_table_s5.csv", sep = ";")
nr5a2_cmps = unique(res_mm[which(res_mm$State == "DOWN"),]$gene_id)

res_mm = read.delim("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/ZGA_Obox_Shuyan/41586_2023_6428_MOESM3_ESM.csv", sep = ";")
OBOX_KO = unique(res_mm[which(res_mm$Group == "down-regulated"),]$gene.name)



df2 <- data.frame(gene=unique(unlist(nVennR_gene_list_polyA_total)))
df1 <- lapply(nVennR_gene_list_polyA_total,function(x){
  data.frame(gene = x)
}) %>% 
  bind_rows(.id = "path")

head(df1)

df_int <- lapply(df2$gene,function(x){
  # pull the name of the intersections
  intersection <- df1 %>% 
    dplyr::filter(gene==x) %>% 
    arrange(path) %>% 
    pull("path") %>% 
    paste0(collapse = "|")
  
  # build the dataframe
  data.frame(gene = x,int = intersection)
}) %>% 
  bind_rows()

library(stringr)
pc = df_int #head(df_int,n=200) 
pc$cnt = stringi::stri_count(pc$int, fixed = "|") +1

head(pc,n=200)


ZGA_genes_all = pc$gene[pc$cnt > 3]

df_int %>% 
  group_by(int) %>% 
  summarise(n=n()) %>% 
  arrange(desc(n))

gene_list_polyA_inter_uni <- list(
  "Nr5a2 KD" = unique(toupper(nr5a2_kd)),
  "Nr5a2 SR1848" = unique(toupper(nr5a2_cmps)),
  "Dux KO" = unique(toupper(Chen_et_al_Dux_targets)),
  "Obox KO" = unique(toupper(OBOX_KO)),
  "ZGA genes" = unique(toupper(ZGA_genes_all)))



library(eulerr)
p <- plot(euler(gene_list_polyA_inter_uni, shape="ellipse"), quantities = TRUE, legend = TRUE)
print(p)


myV <- plotVenn(gene_list_polyA_inter_uni, outFile="/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/gene_list_polyA_inter_uni.svg")


```



```{r}
suppressPackageStartupMessages({
  library(tidyverse)
  library(readxl)
  library(DESeq2)
  library(eulerr)
})

ZGA_Chen_et_al <- file.path("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/Chen_et_al_RNA_seq/41588_2019_418_MOESM3_ESM.xlsx")
ZGA_Chen_et_al <- read_xlsx(ZGA_Chen_et_al, sheet=2)
ZGA_Chen_et_al = ZGA_Chen_et_al[which(ZGA_Chen_et_al$...17 == "down-regulated"),]
Chen_et_al_Dux_targets = ZGA_Chen_et_al[["...1"]] #[-c(1,2)]

res_mm = read.delim("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/science.abn7478_tables_s1_to_s14/science.abn7478_table_s6.csv", sep = ";")
nr5a2_kd = unique(res_mm[which(res_mm$State == "DOWN"),]$gene_id)

res_mm = read.delim("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/science.abn7478_tables_s1_to_s14/science.abn7478_table_s5.csv", sep = ";")
nr5a2_cmps = unique(res_mm[which(res_mm$State == "DOWN"),]$gene_id)

res_mm = read.delim("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/ZGA_Obox_Shuyan/41586_2023_6428_MOESM3_ESM.csv", sep = ";")
OBOX_KO = unique(res_mm[which(res_mm$Group == "down-regulated"),]$gene.name)



df2 <- data.frame(gene=unique(unlist(nVennR_gene_list_polyA)))
df1 <- lapply(nVennR_gene_list_polyA,function(x){
  data.frame(gene = x)
}) %>% 
  bind_rows(.id = "path")

head(df1)

df_int <- lapply(df2$gene,function(x){
  # pull the name of the intersections
  intersection <- df1 %>% 
    dplyr::filter(gene==x) %>% 
    arrange(path) %>% 
    pull("path") %>% 
    paste0(collapse = "|")
  
  # build the dataframe
  data.frame(gene = x,int = intersection)
}) %>% 
  bind_rows()

library(stringr)
pc = df_int #head(df_int,n=200) 
pc$cnt = stringi::stri_count(pc$int, fixed = "|") +1

head(pc,n=200)


ZGA_genes_all = pc$gene[pc$cnt > 4]

df_int %>% 
  group_by(int) %>% 
  summarise(n=n()) %>% 
  arrange(desc(n))

gene_list_polyA_inter_uni <- list(
  "Nr5a2 KD" = unique(toupper(nr5a2_kd)),
  "Nr5a2 SR1848" = unique(toupper(nr5a2_cmps)),
  "Dux KO" = unique(toupper(Chen_et_al_Dux_targets)),
  "Obox KO" = unique(toupper(OBOX_KO)),
  "ZGA genes" = unique(toupper(ZGA_genes_all)))

gene_list_polyA_inter_uni <- list(
  "Nr5a2 KD" = unique(toupper(nr5a2_kd)),
  "Dux KO" = unique(toupper(Chen_et_al_Dux_targets)),
  "Obox KO" = unique(toupper(OBOX_KO)),
  "ZGA genes" = unique(toupper(ZGA_genes_all)))



library(eulerr)
p <- plot(euler(gene_list_polyA_inter_uni, shape="ellipse"), quantities = TRUE, legend = TRUE)
print(p)


myV <- plotVenn(gene_list_polyA_inter_uni, outFile="/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/gene_list_polyA_inter_uni.svg")


```




```{r}
nVennR_gene_list_polyA <- list(
  "C57BL (4)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Hu$SYMBOL)), unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_Yu$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL))))),
  "B6xCAST (1)" = unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
  "CAST/EiJ(mother) x C57BL/6J(father) (1)" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "B6D2F1 (1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
  "C57BL_6NxDBA_2N (1)" = unique(toupper(annotatedRes_Wu$SYMBOL))
  )


nVennR_gene_list_polyA <- list(
  "C57BL/6J clean (N=3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),unique(toupper(annotatedRes_Guo$SYMBOL))))),
  "C57BL/6J x mix (N=3)" = unique(Reduce(intersect,list(unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
                                                 unique(toupper(annotatedRes_Wu$SYMBOL)),
                                              unique(toupper(annotatedRes_Zhang_2016$SYMBOL))))),
  "CAST/EiJxC57BL/6J (N=1)" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "B6D2F1 clean (N=1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
  "ICR clean (N=1)" = unique(toupper(annotatedRes_Yu$SYMBOL))
  )

nVennR_gene_list_polyA <- list(
    "C57BL/6NxPWK/PhJ (1)" = unique(toupper(annotatedRes_Zhang_2016$SYMBOL)),
      "C57BL/6JxCAST/EiJ (1)" = unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
  "C57BL/6NxDBA/2N (1)" = unique(toupper(annotatedRes_Wu$SYMBOL)),
    "ICR (1)" = unique(toupper(annotatedRes_Yu$SYMBOL)),
  "C57BL/6J (3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),unique(toupper(annotatedRes_Guo$SYMBOL)))))
  )


nVennR_gene_list_polyA <- list(
  "C57BL/6J clean (N=3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Shuyan$SYMBOL)), unique(toupper(annotatedRes_C57XBL_6J$SYMBOL)),unique(toupper(annotatedRes_Guo$SYMBOL))))),
  "C57BL/6J x mix (N=3)" = unique(Reduce(intersect,list(unique(toupper(annotatedRes_B6xCAST$SYMBOL)),
                                                 unique(toupper(annotatedRes_Wu$SYMBOL)),
                                              unique(toupper(annotatedRes_Zhang_2016$SYMBOL))))),
  "CAST/EiJxC57BL/6J (N=1)" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "B6D2F1 clean (N=1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
  "ICR clean (N=1)" = unique(toupper(annotatedRes_Yu$SYMBOL))
  )

suppressPackageStartupMessages({
  library(tidyverse)
  library(readxl)
  library(DESeq2)
  library(eulerr)
})

ZGA_Chen_et_al <- file.path("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/Chen_et_al_RNA_seq/41588_2019_418_MOESM3_ESM.xlsx")
ZGA_Chen_et_al <- read_xlsx(ZGA_Chen_et_al, sheet=2)
ZGA_Chen_et_al = ZGA_Chen_et_al[which(ZGA_Chen_et_al$...17 == "down-regulated"),]
Chen_et_al_Dux_targets = ZGA_Chen_et_al[["...1"]] #[-c(1,2)]

res_mm = read.delim("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/science.abn7478_tables_s1_to_s14/science.abn7478_table_s6.csv", sep = ";")
nr5a2_kd = unique(res_mm[which(res_mm$State == "DOWN"),]$gene_id)

res_mm = read.delim("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/science.abn7478_tables_s1_to_s14/science.abn7478_table_s5.csv", sep = ";")
nr5a2_cmps = unique(res_mm[which(res_mm$State == "DOWN"),]$gene_id)

nr5a2_total = unique(Reduce(c,list(unique(toupper(nr5a2_cmps)), unique(nr5a2_kd))))

res_mm = read.delim("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/ZGA_Obox_Shuyan/41586_2023_6428_MOESM3_ESM.csv", sep = ";")
OBOX_KO = unique(res_mm[which(res_mm$Group == "down-regulated"),]$gene.name)



df2 <- data.frame(gene=unique(unlist(nVennR_gene_list_polyA)))
df1 <- lapply(nVennR_gene_list_polyA,function(x){
  data.frame(gene = x)
}) %>% 
  bind_rows(.id = "path")

head(df1)

df_int <- lapply(df2$gene,function(x){
  # pull the name of the intersections
  intersection <- df1 %>% 
    dplyr::filter(gene==x) %>% 
    arrange(path) %>% 
    pull("path") %>% 
    paste0(collapse = "|")
  
  # build the dataframe
  data.frame(gene = x,int = intersection)
}) %>% 
  bind_rows()

library(stringr)
pc = df_int #head(df_int,n=200) 
pc$cnt = stringi::stri_count(pc$int, fixed = "|") +1

head(pc,n=200)


ZGA_genes_all = pc$gene[pc$cnt > 3]

#pc[pc$cnt > 3, ] %>% group_by(int) %>% summarise(n=n())

df_int %>% 
  group_by(int) %>% 
  summarise(n=n()) %>% 
  arrange(desc(n))

# gene_list_polyA_inter_uni <- list(
#   "Nr5a2 KD" = unique(toupper(nr5a2_kd)),
#   "Nr5a2 SR1848" = unique(toupper(nr5a2_cmps)),
#   "Dux KO" = unique(toupper(Chen_et_al_Dux_targets)),
#   "Obox KO" = unique(toupper(OBOX_KO)),
#   "ZGA genes" = unique(toupper(ZGA_genes_all)))


gene_list_polyA_inter_uni <- list(
  "ZGA genes" = unique(toupper(ZGA_genes_all)),
  "NRs regulated" = unique(toupper(nr5a2_total)),
  "Obox regulated" = unique(toupper(OBOX_KO)),
  "Dux regulated" = unique(toupper(Chen_et_al_Dux_targets)))

gene_list_polyA_inter_uni <- list(
  "Dux regulated" = unique(toupper(Chen_et_al_Dux_targets)),
  "Obox regulated" = unique(toupper(OBOX_KO)),
  "NRs regulated" = unique(toupper(nr5a2_total)),
  "ZGA genes" = unique(toupper(ZGA_genes_all)))


# sort = c("ZGA genes","NRs regulated","Obox regulated","Dux regulated")
# 
# gene_list_polyA_inter_uni = names(gene_list_polyA_inter_uni[order(match(gene_list_polyA_inter_uni,sort))])

# gene_list_polyA_inter_uni <- list(
#   "Nr5a2 KD" = unique(toupper(nr5a2_kd)),
#   "Dux KO" = unique(toupper(Chen_et_al_Dux_targets)),
#   "Obox KO" = unique(toupper(OBOX_KO)),
#   "ZGA genes" = unique(toupper(ZGA_genes_all)))



library(eulerr)
p <- plot(euler(gene_list_polyA_inter_uni, shape="ellipse"), quantities = TRUE, legend = TRUE)
print(p)


myV <- plotVenn(nVennR_gene_list_polyA, outFile="/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/gene_list_polyA_inter_uni_nocmpS_unitedNr5a2.svg")


set_size(8, 5)
size = get_size_mode('exclusive_intersection')
ComplexUpset::upset(fromList(gene_list_polyA_inter_uni), 
                    names(gene_list_polyA_inter_uni), 
                    #intersections='all', 
                    width_ratio=0.1, 
                    #min_size=50,
                    #max_size=400,
                    base_annotations = list(
        'Intersection size'=(
            intersection_size(text=list(
                vjust=-0.1,
                hjust=-0.1,
                angle=45,
                color="black"),
                
                text_mapping=aes(
                #label=paste0(
                #    !!upset_text_percentage(),
                #    '\n(', !!size, ')'
                #),
                colour=ifelse(!!size > 50, 'on_bar', 'on_background'),
                y=ifelse(!!size > 50, !!size, !!size)
            )
            
                )
            + ylim(c(0, 5000))
            + theme(axis.text=element_text(size=12),axis.title=element_text(size=14), 
                    axis.line = element_line(colour = "black"),
                    panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),axis.ticks.x=element_line(),axis.ticks.y=element_line()) 
            + ylab('Overlapped genes')
        )
    ),
    #height_ratio=1,
    sort_sets=F,
    set_sizes=(upset_set_size() + theme(axis.text.x=element_text(angle=90))
        + ylab('set size')
        
        ),
    stripes='white', 
    sort_intersections_by=c('degree', 'cardinality')
        ) 

#write.csv(pc[pc$cnt > 3, ], "/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/Pan_ZGA_genes.csv", row.names=FALSE)
#write.csv(pc[pc$cnt == 5, ], "/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/Common_ZGA_genes.csv", row.names=FALSE)

```



```{r}

DEgenes_2cell.vs.G2zygote <- read.csv("/Volumes/pool-totipotency/_from_IMBA/pavelkrav/RNAseq_Proteomics/DEgenes_2cell-vs-G2zygote.csv")
dim(DEgenes_2cell.vs.G2zygote)
DEgenes_2cell.vs.G2zygote_ZGA = DEgenes_2cell.vs.G2zygote[DEgenes_2cell.vs.G2zygote["majZGA"] != "no", ]
dim(DEgenes_2cell.vs.G2zygote_ZGA)
major_ZGA = DEgenes_2cell.vs.G2zygote_ZGA$gene_id

working_dir="/Volumes/pool-totipotency/Pavel/projects/Nr5a2_and_Dux_RNA_seq" 
minor_ZGA <- read.table(file.path(working_dir, "minZGA_genes.csv"), quote="\"", comment.char="")$V1

ZGA_euler_list <- list(
  "ZGA genes Gassler et al. 2022" = unique(toupper(major_ZGA)),
  "Minor ZGA genes Imre" = unique(toupper(minor_ZGA)),
  "Pan ZGA genes" = unique(toupper(pc$gene[pc$cnt > 3])),
  "ZGA genes" = unique(toupper(pc$gene[pc$cnt > 4])))



ZGA_euler_list <- list(
  "ZGA genes Gassler et al. 2022" = unique(toupper(major_ZGA)),
  "Pan ZGA genes" = unique(toupper(pc$gene[pc$cnt > 3])),
  "ZGA genes" = unique(toupper(pc$gene[pc$cnt > 4])))


library(eulerr)
p <- plot(euler(ZGA_euler_list, shape="ellipse"), quantities = TRUE, legend = TRUE)
print(p)


```



```{r}
a = length(unique(Reduce(c,list(
unique(Reduce(intersect,list(unique(toupper(nr5a2_total)), unique(toupper(ZGA_genes_all))))),
unique(Reduce(intersect,list(unique(toupper(Chen_et_al_Dux_targets)), unique(toupper(ZGA_genes_all))))),
unique(Reduce(intersect,list(unique(toupper(OBOX_KO)), unique(toupper(ZGA_genes_all)))))
))))/length(unique(toupper(ZGA_genes_all)))
b = 1-a

gene_list_polyA_inter_uni <- list(
  "NOD regulated" = a,
  "Others" = b)



# Load ggplot2
library(ggplot2)

# Create Data
data <- data.frame(
  group=names(gene_list_polyA_inter_uni),
  value=unlist(gene_list_polyA_inter_uni)*100
)

data <- data %>% 
  arrange(desc(group)) %>%
  mutate(prop = value / sum(data$value) *100) %>%
  mutate(ypos = cumsum(prop)- 0.5*prop )

# Basic piechart
ggplot(data, aes(x="", y=value, fill=group)) +
geom_bar(stat="identity", width=1, color="black") +
  coord_polar("y", start=0) +
  theme_void() + 
  theme(legend.position="none") +
  geom_text(aes(y = ypos, label = group), color = "black", size=6) +
  scale_fill_brewer(palette="Set1")




library(ggplot2)
library(webr)

# building the pie-donut chart
PieDonut(data, aes(group, count=value),
         title = "Browser market share, April, 2011",
         ratioByGroup = FALSE)

```



```{r}
# from https://gist.github.com/mtmorgan/eaf456ad5b45c431c494#file-readkallisto-r
# source("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/readKallisto.R")
# files <- dir("/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko/ZGA_Wang", "abundance.tsv", full=TRUE,
#              recursive=TRUE)
# stopifnot(all(file.exists(files)))
# files <- sub(".tsv", ".h5", files, fixed=TRUE)
# exp = readKallisto(files, what=c("tpm", "eff_length"), as="SummarizedExperiment")
# exp@assays@data$tpm
# 
# 
# str(readKallisto(files, as="matrix"))
# str(readKallisto(files, as="list"))
# (se <- readKallisto(files, as="SummarizedExperiment"))
# 
# str(readKallisto(files, what="eff_length"))
# readKallisto(files, what="eff_length", as="SummarizedExperiment")
# 
# files <- sub(".tsv", ".h5", files, fixed=TRUE)
# exp = readKallisto(files, what=c("tpm", "eff_length"), as="SummarizedExperiment")
# exp@assays@data$tpm
# 
# 
# 
# str(readKallisto(files))
# str(readKallisto(files, what="tpm"))
# readKallisto(files, what="tpm", as="SummarizedExperiment")
# readKallisto(files, what=c("tpm", "eff_length"), as="SummarizedExperiment")
# 
# xx <- readBootstrap(files[1])
# ridx <- head(which(rowSums(xx) != 0), 3)
# cidx <- c(1:5, 96:100)
# readBootstrap(files[1])[ridx, cidx]
# readBootstrap(files[1], i=c(ridx, rev(ridx)), j=cidx)
# 
# 
# exp = readKallisto(files, what=c("tpm", "eff_length"), as="SummarizedExperiment")
# exp@assays@data$tpm
```
