---
title: "10092023_ZGA_List_intersection"
output:
  html_document:
    df_print: paged
  pdf_document: default
---

# Title
# Date 10092023
# Name ZGA_List_intersection
# Description The R script for DE analysis and Pan ZGA genes list reproduction


## Loading libraries
```{r}
library(conflicted)
library(rhdf5)
library(org.Mm.eg.db)
library(tximport)
library(DESeq2)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(magrittr)
library(eulerr)
library(nVennR)
library(UpSetR)
library(ComplexUpset)
library(forcats)
library(webr)
library(readxl)
library(RColorBrewer)
library(stringr)

"%+%" <- function(...){
  paste0(...)
}

source("./readKallisto.R") # from Martin Morgan (mtmorgan)
#source(textConnection(getURL('https://gist.githubusercontent.com/mtmorgan/eaf456ad5b45c431c494/raw/2700e5002a0145ab00bf32bdb07edd2052c4c842/readKallisto.R')))
# system("wget https://github.com/pachterlab/kallisto-transcriptome-indices/releases/download/ensembl-96/mus_musculus.tar.gz")
# system("tar -xf mus_musculus.tar.gz")
```


# Prepearing to process RNA-Seq
```{r}

# processRNA_Seq function: from a path to Kallisto files to a sumary DE table and volcano plot

processRNA_Seq <- function(path, directory, gtf.gr, plot_volcanoplot=T) {
  
  # solving some conflicts
  conflict_prefer("mutate", "dplyr")
  conflict_prefer("strsplit", "Biostrings")
  conflict_prefer("getSequence", "biomaRt")
  conflict_prefer("select", "dplyr")
  
  # setting up environment
  working_dir = path %+% "/" %+% directory
  
  # loading methadata
  samples_methadata <- read.csv(file.path(working_dir, "samples.csv"), header = TRUE, sep=";")
  
  # loading annotation
  gtf.df = as.data.frame(gtf.gr)
  txt2gene=(gtf.df[gtf.df$type=='transcript' ,c("transcript_id","gene_name")])
  names(txt2gene)<-c('TXNAME','GENEID')
  
  # loading kallisto h5
  h5closeAll() # close to avoid conflicts
  kallisto_files <- file.path(working_dir, "Kallisto", samples_methadata$file_path, "abundance.h5")
  names(kallisto_files) <- samples_methadata$file_path
  txi.kallisto <- tximport(kallisto_files, type = "kallisto", txOut = F, tx2gene = txt2gene, ignoreTxVersion = TRUE)
  
  write.table(txi.kallisto$counts,file=paste0(working_dir, "/txi.kallisto_counts_raw_genes.csv"),row.names=T,quote=F,sep="\t")
  
  # running DESeq2 
  dds <- DESeqDataSetFromTximport(txi.kallisto, colData = samples_methadata, design = ~condition)
  dds <- DESeq(dds)
  fpkmdds = as.data.frame(fpkm(dds))
  fpkmdds$Zygote = apply(fpkmdds[, samples_methadata$sample_name[samples_methadata["condition"] == "Zygote"]],1,mean)
  fpkmdds$`2cell` = apply(fpkmdds[, samples_methadata$sample_name[samples_methadata["condition"] == "2cell"]],1,mean)
  fpkmdds = fpkmdds %>% select(c("Zygote", "2cell"))
  colnames(fpkmdds) = c("Zygote_FPKM", "2cell_FPKM")
  abdds = as.data.frame(txi.kallisto$abundance)
  abdds$Zygote = apply(abdds[, samples_methadata$sample_name[samples_methadata["condition"] == "Zygote"]],1,mean)
  abdds$`2cell` = apply(abdds[, samples_methadata$sample_name[samples_methadata["condition"] == "2cell"]],1,mean)
  abdds = abdds %>% dplyr::select(c("Zygote", "2cell"))
  colnames(abdds) = c("Zygote_Ab", "2cell_Ab")
  
  # importing abundances
  files <- dir(working_dir, "abundance.tsv", full=TRUE,
               recursive=TRUE)
  stopifnot(all(file.exists(files)))
  files <- sub(".tsv", ".h5", files, fixed=TRUE)
  exp = readKallisto(files, what=c("tpm", "eff_length"), as="SummarizedExperiment")
  tpmdds = as.data.frame(exp@assays@data$tpm)
  tpmdds$Zygote = apply(tpmdds[, samples_methadata$sample_name[samples_methadata["condition"] == "Zygote"]],1,mean)
  tpmdds$`2cell` = apply(tpmdds[, samples_methadata$sample_name[samples_methadata["condition"] == "2cell"]],1,mean)
  tpmdds = tpmdds %>% dplyr::select(c("Zygote", "2cell"))
  colnames(tpmdds) = c("Zygote_TPM", "2cell_TPM")
  tpmdds$ENTREZID = rownames(tpmdds) 
  tpmdds = tidyr::separate(tpmdds, col="ENTREZID", into = c("ENTREZID", "ENTREZID_dot")) %>% dplyr::select(c("ENTREZID", "Zygote_TPM", "2cell_TPM"))
  
  # merging dfs to form an output
  tpmdds <- merge(gtf.df %>% dplyr::select(c("transcript_id", "gene_name")),tpmdds,
                        by.x=1,
                        by.y=1,
                        all.x=FALSE,
                        all.y=TRUE)
  
  tpmdds = tpmdds[!duplicated(tpmdds), ]
  myResS <- results(dds, contrast = c("condition", "Zygote", "2cell"))
  myResS <- myResS[order(myResS$pvalue), ]
  myResAsDF <- as.data.frame(myResS)
  myResAsDF$newPadj <- p.adjust(myResAsDF$pvalue, method ="fdr")
  
  conflict_prefer("select", "AnnotationDbi")
  eToSym <- select(org.Mm.eg.db,
                   keys = rownames(myResAsDF),
                   keytype = "SYMBOL",
                   columns=c("SYMBOL","ENTREZID","GENENAME"))
  
  annotatedRes <- merge(eToSym,myResAsDF,
                        by.x=1,
                        by.y=0,
                        all.x=TRUE,
                        all.y=TRUE)
  
  annotatedRes <- annotatedRes[order(annotatedRes$pvalue),]
  
  annotatedRes_fpkmdds <- merge(annotatedRes,fpkmdds,
                        by.x=1,
                        by.y=0,
                        all.x=TRUE,
                        all.y=FALSE)
  
  annotatedRes_fpkmdds <- merge(annotatedRes_fpkmdds,tpmdds,
                        by.x=1,
                        by.y=2,
                        all.x=TRUE,
                        all.y=FALSE)
  
  write.table(annotatedRes_fpkmdds,file=paste0("./tables/annotatedRes_log2FC_" %+% directory %+% "_genes.csv"),row.names=T,quote=F,sep="\t")
  
  if (plot_volcanoplot == T) {
    
    # Visualizing DE genes on volcano plot
    annotatedRes %>% 
      mutate(padj_neg10 = -log10(padj)) -> res_mm
    
    FLAG = 4
    groupB_criteria <- res_mm$padj < 0.01 & abs(res_mm$log2FoldChange) > FLAG
    
    res_mm$significance <- "n.s."
    group1 = "padj < 0.01"
    res_mm$significance[res_mm$padj < 0.01] <- group1
    group2 = "padj < 0.01 & log2FoldChange > " %+% FLAG
    res_mm$significance[groupB_criteria] <- group2
    
    ggplot(res_mm, aes(log2FoldChange, padj_neg10, color = significance)) +
      geom_point(shape = 16, alpha = 0.6) +
      labs(title = "DESeq2 transcripts " %+% "Zygote_G2" %+% " vs " %+% "2cell_G2") +
      xlab("log2FoldChange")+ ylab("-log10(padj)")+
      guides(text = guide_legend()) +
      scale_color_manual(values = c("n.s." = "darkgray",
                                    "padj < 0.01" = "gray",
                                    "padj < 0.01 & log2FoldChange > 4" = "orange")) +
      NULL + theme_light()
    
    ggsave("./figures/DE_"%+% directory %+%".pdf")
  }
  return(txi.kallisto$counts)
}
```


# Processing RNA-Seq
```{r}
home = "/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko"

gtf.gr = rtracklayer::import("./mus_musculus/Mus_musculus.GRCm38.96.gtf")

cnt1 = processRNA_Seq(home, "ZGA_Gassler_et_al_C57BL_6J_GSE178638_paired_polyA", gtf.gr)
cnt2 = processRNA_Seq(home, "ZGA_Gassler_et_al_B6xCAST_GSE178638_paired_polyA", gtf.gr)
cnt3 = processRNA_Seq(home, "ZGA_Ji_et_al_C57BL_6_GSE215813_paired_polyA", gtf.gr)
cnt4 = processRNA_Seq(home, "ZGA_Wu_et_al_C57BL_6NxDBA_2N_GSE66582_paired_polyA", gtf.gr)
cnt5 = processRNA_Seq(home, "ZGA_Chen_et_al_B6D2F1_GSE121746_single_polyA", gtf.gr)
cnt6 = processRNA_Seq(home, "ZGA_Deng_et_al_CAST_EiJxC57BL_6J_GSE45719_single_polyA", gtf.gr)
cnt7 = processRNA_Seq(home, "ZGA_Guo_et_al_C57BL_6_GSE134832_paired_polyA", gtf.gr)
cnt8 = processRNA_Seq(home, "ZGA_Zhang_et_al_C57BL_6NxPWK_PhJ_GSE71434_paired_polyA", gtf.gr)
```

# Checking integrity
```{r}

df_vector = list(cnt1, cnt2, cnt3, cnt4, cnt5, cnt6, cnt7, cnt8) 
names_vector = c(rep("Gassler_et_al_C57BL_6J", ncol(cnt1)), 
                 rep("Gassler_et_al_B6xCAST", ncol(cnt2)), 
                 rep("Ji_et_al_C57BL_6", ncol(cnt3)),
                 rep("Wu_et_al_C57BL_6NxDBA_2N", ncol(cnt4)), 
                 rep("Chen_et_al_B6D2F1", ncol(cnt5)), 
                 rep("Deng_et_al_CAST_EiJxC57BL_6J", ncol(cnt6)), # IVF dataset
                 rep("Guo_et_al_C57BL_6", ncol(cnt7)), 
                 rep("Zhang_et_al_C57BL_6NxPWK_PhJ", ncol(cnt8)))


normalized_list <- lapply(df_vector, function(df) {
  df[, -1] <- scale(df[, -1])
  return(df)
})


combined_dataframe <- do.call(cbind, normalized_list)

umap_result <- umap::umap(t(combined_dataframe))

umap_df <- data.frame(UMAP1 = umap_result$layout[, 1],
                      UMAP2 = umap_result$layout[, 2],
                      Dataset = names_vector)

n <- length(df_vector)
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))

ggplot(umap_df, aes(x = UMAP1, y = UMAP2, color=Dataset)) +
  geom_point(size=5, alpha=0.5) +
  labs(title = "UMAP Plot Colored by Dataset") +
  theme_minimal()+
  scale_color_manual(values = col_vector)
ggsave("./figures/UMAP_datasets.pdf")
```

# Filtering ZGA genes
```{r}
conflict_prefer("select", "dplyr")

v_1cell_threshold = 2
v_2cell_threshold = 5
v_pval_threshold = 0.05
log2FoldChange_threshold = 2
home = "/Volumes/pool-toti-bioinfo/bioinfo/Pavel.Kravchenko"
tables = "/Users/pavel/Desktop/Rise_and_SINE/tables"

read_and_threshold_ZGA_genes <- function(path, directory, v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold){
  working_dir = path #%+% "/" %+% directory
  annotatedRes = read.csv(file=paste0(working_dir, "/annotatedRes_log2FC_" %+% directory %+% "_genes.csv"),sep="\t")
  annotatedRes %<>% dplyr::filter((log2FoldChange < log2FoldChange_threshold) & (newPadj < v_pval_threshold) & (Zygote_FPKM > v_1cell_threshold) & (X2cell_FPKM > v_2cell_threshold))
  return(annotatedRes)
}

annotatedRes_Gassler_C57XBL_6J = read_and_threshold_ZGA_genes(tables, "ZGA_Gassler_et_al_C57BL_6J_GSE178638_paired_polyA", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_Gassler_B6xCAST = read_and_threshold_ZGA_genes(tables, "ZGA_Gassler_et_al_B6xCAST_GSE178638_paired_polyA", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_Ji = read_and_threshold_ZGA_genes(tables, "ZGA_Ji_et_al_C57BL_6_GSE215813_paired_polyA", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_Wu = read_and_threshold_ZGA_genes(tables, "ZGA_Wu_et_al_C57BL_6NxDBA_2N_GSE66582_paired_polyA", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_Chen = read_and_threshold_ZGA_genes(tables, "ZGA_Chen_et_al_B6D2F1_GSE121746_single_polyA", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_Deng = read_and_threshold_ZGA_genes(tables, "ZGA_Deng_et_al_CAST_EiJxC57BL_6J_GSE45719_single_polyA", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_Guo = read_and_threshold_ZGA_genes(tables, "ZGA_Guo_et_al_C57BL_6_GSE134832_paired_polyA", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
annotatedRes_Zhang = read_and_threshold_ZGA_genes(tables, "ZGA_Zhang_et_al_C57BL_6NxPWK_PhJ_GSE71434_paired_polyA", v_1cell_threshold, v_2cell_threshold, v_pval_threshold, log2FoldChange_threshold)
```


# Overlapping ZGA lists for the mouse strains
```{r}
gene_list <- list(
  "B6xCAST Gassler" = unique(toupper(annotatedRes_Gassler_B6xCAST$SYMBOL)),
  "C57BL_6J Gassler" = unique(toupper(annotatedRes_Gassler_C57XBL_6J$SYMBOL)),
  "C57BL_6N Ji" = unique(toupper(annotatedRes_Ji$SYMBOL)),
  "C57BL_6NxDBA_2N Wu" = unique(toupper(annotatedRes_Wu$SYMBOL)),
  "B6D2F1 Chen" = unique(toupper(annotatedRes_Chen$SYMBOL)),
  "CAST_EiJxC57BL_6J Deng" = unique(toupper(annotatedRes_Deng$SYMBOL)),
  "B6D2F1_J Zhang" = unique(toupper(annotatedRes_Zhang$SYMBOL)),
  "C57BL_6 Guo" = unique(toupper(annotatedRes_Guo$SYMBOL)))

conflicts_prefer(base::intersect)
nVennR_gene_list_polyA <- list(
     "B6D2F1 pure (N=1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
     "C57BL/6 x mix (N=3)" = unique(Reduce(intersect,list(unique(toupper(annotatedRes_Gassler_B6xCAST$SYMBOL)),
                                              unique(toupper(annotatedRes_Wu$SYMBOL)),
                                              unique(toupper(annotatedRes_Zhang$SYMBOL))))),
     "C57BL/6 pure (N=3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Ji$SYMBOL)),
                                          unique(toupper(annotatedRes_Gassler_C57XBL_6J$SYMBOL)),
                                          unique(toupper(annotatedRes_Guo$SYMBOL))))),
  "CAST/EiJxC57BL/6J (N=1)" = unique(toupper(annotatedRes_Deng$SYMBOL))
  )

pdf("./figures/nVennR_gene_list_polyA.pdf")
p = plot(euler(nVennR_gene_list_polyA, shape="ellipse"), quantities = TRUE, legend = TRUE)
print(p)
dev.off()
pV = plotVenn(nVennR_gene_list_polyA, outFile="./figures/nVennR_plot_polyA.svg")

set_size = function(w, h, factor=1.5) {
    s = 1 * factor
    options(
        repr.plot.width=w * s,
        repr.plot.height=h * s,
        repr.plot.res=100 / factor,
        jupyter.plot_mimetypes='image/png',
        jupyter.plot_scale=1
    )
}

pdf("./figures/gene_list_polyA_UpSetR_plot.pdf")
set_size(8, 5)
size = get_size_mode('exclusive_intersection')
ComplexUpset::upset(fromList(nVennR_gene_list_polyA), #[apply(fromList(nVennR_gene_list_polyA_total),1,sum) > 3, ], 
                    names(nVennR_gene_list_polyA), 
                    width_ratio=0.1, 
                    base_annotations = list(
        'Intersection size'=(
            intersection_size(text=list(
                vjust=-0.1,
                hjust=-0.1,
                angle=45,
                color="black"),
                text_mapping=aes(
                colour=ifelse(!!size > 50, 'on_bar', 'on_background'),
                y=ifelse(!!size > 50, !!size, !!size)))
            + ylim(c(0, 1750))
            + theme(axis.text=element_text(size=12),
                    axis.title=element_text(size=14), 
                    axis.line = element_line(colour = "black"),
                    panel.grid.major = element_blank(),
                    panel.grid.minor = element_blank(),
                    axis.ticks.x=element_line(),
                    axis.ticks.y=element_line()) 
            + ylab('Overlapped genes')
        )
    ),
    sort_sets='descending',
    set_sizes=(upset_set_size() + theme(axis.text.x=element_text(angle=90))
        + ylab('set size')
        ),
    stripes='white', 
    sort_intersections_by=c('degree', 'cardinality')
        ) 
dev.off()
```


# Selectiong Pan ZGA list
$$
Pan \ ZGA \ list = \{ \bigcap_{\alpha = 1}^{ð”¸} Z \alpha\} \bigcup_{}^{} \{\bigcup_{k= 1}^{ð”¸}\{\bigcap_{ \alpha \neq k \atop \forall \alpha âˆˆð”¸} Z \alpha\}\}, \ where \ ð”¸ \ is \ a \ set \ of \ ZGA \ datasets \ Z1, Z2, ... ,Zn
$$


# Checking Pan categories
```{r}
cc = c(nrow(fromList(gene_list)[apply(fromList(gene_list),1,sum) == 8, ]))

for (i in seq(length(gene_list))){
  print(i)
  cc = c(cc, nrow(fromList(gene_list[- i])[apply(fromList(gene_list[- i]),1,sum) > 6, ]))
}

cn = c("All", paste0("-",names(gene_list)))
df = as.data.frame(list(Overlap=cc,Datasets=cn))

range = max(df$Overlap) -min(df$Overlap)

ggplot(df, aes(x=fct_inorder(cn), y=cc)) +
geom_bar(stat="identity",colour=NA,linewidth=0) +
  theme_classic() +
  ylab("Overlaped genes") +
  xlab("Datasets")+theme(axis.text.x = element_text(angle = 90),
     axis.line = element_line(colour = "black"),
     text = element_text(size = 12),
     panel.grid.major = element_blank(),
  panel.grid.minor = element_blank()) +
  geom_text(aes(label = cc), 
            angle = 45,
            nudge_y = range/3) +
  ggtitle("ZGA datasets intersection. \n All categories and all - k")

ggsave("./figures/Pan_ZGA_gene_list_barplot.pdf")
```



# Checking Pan categories
```{r}
cc = c(nrow(fromList(nVennR_gene_list_polyA)[apply(fromList(nVennR_gene_list_polyA),1,sum) == 4, ]))

for (i in seq(length(nVennR_gene_list_polyA))){
  print(i)
  cc = c(cc, nrow(fromList(nVennR_gene_list_polyA[- i])[apply(fromList(nVennR_gene_list_polyA[- i]),1,sum) > 2, ]))
}

cn = c("All", paste0("-",names(nVennR_gene_list_polyA)))
df = as.data.frame(list(Overlap=cc,Datasets=cn))

range = max(df$Overlap) -min(df$Overlap)

ggplot(df, aes(x=fct_inorder(cn), y=cc)) +
geom_bar(stat="identity",colour=NA,linewidth=0) +
  theme_classic()+
  ylab("Overlaped genes") +
  xlab("Datasets")+theme(axis.text.x = element_text(angle = 90),
     axis.line = element_line(colour = "black"),
     text = element_text(size = 12),
     panel.grid.major = element_blank(),
  panel.grid.minor = element_blank()) +
  geom_text(aes(label = cc), 
            angle = 0,
            nudge_y = range/3) +
  ggtitle("Strain specific ZGA genes intersection. \n All categories and all - k")

ggsave("./figures/Strain_specific_Pan_ZGA_gene_list_polyA_barplot.pdf")
```


```{r}
nVennR_gene_list_polyA <- list(
     "B6D2F1 pure (N=1)" = unique(toupper(annotatedRes_Chen$SYMBOL)),
     "C57BL/6 x mix (N=3)" = unique(Reduce(intersect,list(unique(toupper(annotatedRes_Gassler_B6xCAST$SYMBOL)),
                                              unique(toupper(annotatedRes_Wu$SYMBOL)),
                                              unique(toupper(annotatedRes_Zhang$SYMBOL))))),
     "C57BL/6 pure (N=3)" = unique(Reduce(c,list(unique(toupper(annotatedRes_Ji$SYMBOL)),
                                          unique(toupper(annotatedRes_Gassler_C57XBL_6J$SYMBOL)),
                                          unique(toupper(annotatedRes_Guo$SYMBOL))))),
  "CAST/EiJxC57BL/6J (N=1)" = unique(toupper(annotatedRes_Deng$SYMBOL))
  )

Dux_KO <- file.path("./tables/41588_2019_418_MOESM3_ESM.xlsx")
Dux_KO <- read_xlsx(Dux_KO, sheet=2)
Dux_KO = Dux_KO[which(Dux_KO$...17 == "down-regulated"),]
Dux_KO = unique(toupper(na.omit(Dux_KO[["...1"]]))) #[-c(1,2)]
write.csv(Dux_KO, "./tables/Dux_KO_ZGA_genes.csv", row.names=FALSE)

res_mm = read.delim("./tables/science.abn7478_table_s6.csv", sep = ";")
Nr5a2_KD = unique(toupper(na.omit(unique(res_mm[which(res_mm$State == "DOWN"),]$gene_id))))
write.csv(Nr5a2_KD, "./tables/Nr5a2_KD_ZGA_genes.csv", row.names=FALSE)

res_mm = read.delim("./tables/science.abn7478_table_s5.csv", sep = ";")
Nr5a2_SR1848 = unique(toupper(na.omit(unique(res_mm[which(res_mm$State == "DOWN"),]$gene_id))))
write.csv(Nr5a2_SR1848, "./tables/Nr5a2_SR1848_ZGA_genes.csv", row.names=FALSE)

Nr5a2_total = unique(toupper(na.omit(unique(Reduce(c,list(unique(toupper(Nr5a2_SR1848)), unique(Nr5a2_KD)))))))
write.csv(Nr5a2_total, "./tables/Nr5a2_total_ZGA_genes.csv", row.names=FALSE)

res_mm = read.delim("./tables/41586_2023_6428_MOESM3_ESM.csv", sep = ";")
Obox_KO = unique(toupper(na.omit(unique(res_mm[which(res_mm$Group == "down-regulated"),]$gene.name))))
write.csv(Obox_KO, "./tables/Obox_KO_ZGA_genes.csv", row.names=FALSE)


df2 <- data.frame(gene=unique(unlist(nVennR_gene_list_polyA)))
write.csv(df2, "./ZGA_genes_summary_table.csv", row.names=FALSE)

df1 <- lapply(nVennR_gene_list_polyA,function(x){
  data.frame(gene = x)
}) %>% 
  bind_rows(.id = "path")

df_int <- lapply(df2$gene,function(x){
  intersection <- df1 %>% 
    dplyr::filter(gene==x) %>% 
    arrange(path) %>% 
    pull("path") %>% 
    paste0(collapse = "|")
  
  data.frame(gene = x,int = intersection)
}) %>% 
  bind_rows()


pc = df_int
pc$cnt = stringi::stri_count(pc$int, fixed = "|") +1

ZGA_genes_all = pc$gene[pc$cnt > 2]

df_int %>% 
  group_by(int) %>% 
  summarise(n=n()) %>% 
  arrange(desc(n))

gene_list_polyA_inter_uni <- list(
  "Dux regulated" = unique(toupper(Dux_KO)),
  "Obox regulated" = unique(toupper(Obox_KO)),
  "NRs regulated" = unique(toupper(Nr5a2_total)),
  "ZGA genes" = unique(toupper(ZGA_genes_all)))


p = plot(euler(gene_list_polyA_inter_uni, shape="ellipse"), quantities = TRUE, legend = TRUE)
print(p)

# calculating individual overlaps
# length(unique(Reduce(intersect,list(unique(toupper(Dux_KO)), unique(toupper(ZGA_genes_all)))))) #34
# length(unique(Reduce(intersect,list(unique(toupper(Obox_KO)), unique(toupper(ZGA_genes_all)))))) #404
# length(unique(Reduce(intersect,list(unique(toupper(Nr5a2_total)), unique(toupper(ZGA_genes_all)))))) #927
# length(unique(Reduce(intersect,list(unique(toupper(Nr5a2_KD)), unique(toupper(ZGA_genes_all)))))) #122
# length(unique(Reduce(intersect,list(unique(toupper(Nr5a2_SR1848)), unique(toupper(ZGA_genes_all)))))) #122

pdf("./figures/NRs_Dux_Obox_Pan_ZGA_gene_list_polyA_euler.pdf")
p = plot(euler(gene_list_polyA_inter_uni, shape="ellipse"), quantities = TRUE, legend = TRUE)
print(p)
dev.off()
pV = plotVenn(nVennR_gene_list_polyA, outFile="./figures/nVennR_NRs_Dux_Obox_Pan_ZGA_gene_list_polyA.svg")


pdf("./figures/NRs_Dux_Obox_Pan_ZGA_gene_list_polyA_UpSetR_plot_1500.pdf")
set_size(8, 5)
size = get_size_mode('exclusive_intersection')
ComplexUpset::upset(fromList(gene_list_polyA_inter_uni), 
                    names(gene_list_polyA_inter_uni), 
                    width_ratio=0.1,
                    base_annotations = list(
        'Intersection size'=(
            intersection_size(text=list(
                vjust=-0.1,
                hjust=-0.1,
                angle=45,
                color="black"),
                mapping=aes(fill='bars_color', colour = "black"),
                text_mapping=aes(
                colour=ifelse(!!size > 50, 'on_bar', 'on_background'),
                y=ifelse(!!size > 50, !!size, !!size)
                )
              )
            + scale_fill_manual(values=c('bars_color'='white'), guide='none')
            + ylim(c(0, 1500))
            + theme(axis.text=element_text(size=12),
                    axis.title=element_text(size=14), 
                    axis.line = element_line(colour = "black"),
                    panel.grid.major = element_blank(),
                    panel.grid.minor = element_blank(),
                    axis.ticks.x=element_line(),
                    axis.ticks.y=element_line()) 
            + ylab('Overlapped genes')
        )
    ),
    sort_sets=F,
    set_sizes=(upset_set_size() + theme(axis.text.x=element_text(angle=90))
        + ylab('set size')
        ),
    stripes='white', 
    sort_intersections_by=c('degree', 'cardinality')
        ) 
dev.off()


pdf("./figures/NRs_Dux_Obox_Pan_ZGA_gene_list_polyA_UpSetR_plot_5000.pdf")
set_size(8, 5)
size = get_size_mode('exclusive_intersection')
ComplexUpset::upset(fromList(gene_list_polyA_inter_uni), 
                    names(gene_list_polyA_inter_uni), 
                    width_ratio=0.1,
                    base_annotations = list(
        'Intersection size'=(
            intersection_size(text=list(
                vjust=-0.1,
                hjust=-0.1,
                angle=45,
                color="black"),
                #mapping=aes(fill=F),
                mapping=aes(fill='bars_color', colour = "black"),
                text_mapping=aes(
                colour=ifelse(!!size > 50, 'on_bar', 'on_background'),
                y=ifelse(!!size > 50, !!size, !!size)
                )
              )
            + scale_fill_manual(values=c('bars_color'='white'), guide='none')
            + ylim(c(0, 5000))
            + theme(axis.text=element_text(size=12),
                    axis.title=element_text(size=14), 
                    axis.line = element_line(colour = "black"),
                    panel.grid.major = element_blank(),
                    panel.grid.minor = element_blank(),
                    axis.ticks.x=element_line(),
                    axis.ticks.y=element_line()) 
            + ylab('Overlapped genes')
        )
    ),
    sort_sets=F,
    set_sizes=(upset_set_size() + theme(axis.text.x=element_text(angle=90))
        + ylab('set size')
        ),
    stripes='white', 
    sort_intersections_by=c('degree', 'cardinality')
    

        ) 
dev.off()

colnames(pc) = c("Gene_id", "Intersection_category", "Intersection_count")
write.csv(pc[pc$Intersection_count > 2, ], "./Pan_ZGA_lists/Pan_1_degree_ZGA_genes.csv", row.names=FALSE)
write.csv(pc[pc$Intersection_count > 1, ], "./Pan_ZGA_lists/Pan_2_degree_ZGA_genes.csv", row.names=FALSE)
write.csv(pc[pc$Intersection_count > 0, ], "./Pan_ZGA_lists/Pan_3_degree_ZGA_genes.csv", row.names=FALSE)
write.csv(pc[pc$Intersection_count == 4, ], "./Pan_ZGA_lists/Core_ZGA_genes.csv", row.names=FALSE)
write.csv(pc, "./Pan_ZGA_lists/Complete_ZGA_genes.csv", row.names=FALSE)

```


```{r}
a = length(unique(Reduce(c,list(
unique(Reduce(intersect,list(unique(toupper(Nr5a2_total)), unique(toupper(ZGA_genes_all))))),
unique(Reduce(intersect,list(unique(toupper(Dux_KO)), unique(toupper(ZGA_genes_all))))),
unique(Reduce(intersect,list(unique(toupper(Obox_KO)), unique(toupper(ZGA_genes_all)))))
))))/length(unique(toupper(ZGA_genes_all)))

b = 1-a

gene_list_polyA_inter_uni <- list(
  "NOD regulated" = a,
  "Others" = b)

data <- data.frame(
  group=names(gene_list_polyA_inter_uni),
  value=unlist(gene_list_polyA_inter_uni)*100
)

data <- data %>% 
  arrange(desc(group)) %>%
  mutate(prop = value / sum(data$value) *100) %>%
  mutate(ypos = cumsum(prop)- 0.5*prop )

pdf("./figures/ZGA_genes_regulation_pie_chart.pdf")
PieDonut(data, aes(group, count=value),
         ratioByGroup = FALSE)
dev.off()
```


```{r}
sessionInfo()
```